lib/stream.c

```diff
@@ -100,16 +100,12 @@ struct stream *stream_new(size_t size)
 
 	assert(size > 0);
 
-	s = XCALLOC(MTYPE_STREAM, sizeof(struct stream));
+	s = XMALLOC(MTYPE_STREAM, sizeof(struct stream));
 
-	if (s == NULL)
```

Comment:
Are we adopting the notion that `malloc`/`calloc` should never fail on resource exhaustion?

For normal or big machines I wouldn't see this as a problem (e.g. like most of modern programming languages), but for embedded devices it might be desirable to see an attempt to avoid a crash.

Example of situation of embedded devices which might want this: you have a low ram device running a routing protocol and you run some tools along with it, say a rrd tool for graphs. Eventually you'll find a period where the RAM usage might spike and you'll have a resource exhaustion situation because the protocol wants to converge/reconverge and the rrd tool wants to make a new graph.

---

I ask this question because I think that not handling memory exhaustion is much easier than doing it. If we embrace this strategy we might simplify other parts of the code and avoid the cost of conditional branch on each memory allocation (speed!), but sacrifice resource exhaustion control on embedded and restricted environments (e.g. *BSDs might hit the memory limit quicker).